conda_forge_tick/migrators.xsh

```diff
@@ -0,0 +1,231 @@
+    delim = ["'", '"']
+    sets = [' set', 'set']
+    base1 = '''{{%{set} {checkname} = {d}[0-9A-Fa-f]+{d} %}}'''
+    base2 = '''{{%{set} {checkname} = {d}$HASH{d} %}}'''
+    for cn in checksum_names:
+        for s in sets:
+            for d in delim:
+                more_patterns.append(('meta.yaml',
+                                      base1.format(set=s, checkname=cn, d=d),
+                                      base2.format(set=s, checkname=cn, d=d)))
+
+    def filter(self, attrs):
+        conditional = super().filter(attrs)
+        return bool(conditional  # if archived
+                or not attrs.get('new_version')  # if no new version
+                # if new version is less than current version
+                or parse_version(str(attrs['new_version'])) <= parse_version(str(attrs['version']))
+                # if PRed version is greater than newest version
+                or attrs.get('PRed', '0.0.0') >= parse_version(attrs['new_version']))
+
+    def migrate(self, recipe_dir, attrs, hash_type='sha256'):
+        # Render with new version but nothing else
+        with indir(recipe_dir):
+            for f, p, n in patterns:
+                p = eval_version(p)
+                n = eval_version(n)
+                replace_in_file(p, n, f)
+            with open('meta.yaml', 'r') as f:
+                text = f.read()
+            # If we can't parse the meta_yaml then jump out
+            meta_yaml = parsed_meta_yaml(text)
+            # If the parser returns None, then we didn't read the meta.yaml
+            # TODO: How we didn't fail at 01 on this recipe is mysterious
+            if meta_yaml is None:
+                attrs['bad'] = '{}: failed to read meta.yaml\n'.format($PROJECT)
+                return False
+            source_url = meta_yaml.get('source', {}).get('url')
+            if not source_url:
+                attrs['bad'] = '{}: missing url\n'.format($PROJECT)
+                return False
+            if isinstance(source_url, list):
+                for url in source_url:
+                    if 'Archive' not in url:
+                        source_url = url
+                        break
+            if 'cran.r-project.org/src/contrib' in source_url:
+                $VERSION = $VERSION.replace('_', '-')
+
+        # now, update the feedstock to the new version
+        source_url = eval_version(source_url)
+        try:
+            hash = hash_url(source_url, hash_type)
+        except urllib.error.HTTPError:
+            attrs['bad'] = '{}: hash failed at {}\n'.format(
+                meta_yaml.get('package', {}).get('name', 'UNKOWN'), source_url)
+            return False
+
+        patterns += tuple(more_patterns)
+        with indir(recipe_dir), ${...}.swap(HASH_TYPE=hash_type, HASH=hash, SOURCE_URL=source_url):
+            for f, p, n in patterns:
+                p = eval_version(p)
+                n = eval_version(n)
+                replace_in_file(p, n, f)
+        return True
```

Comment:
Why is this returning True, if the Migrator API says it should return None?

Reply:
Sorry it should return a bool, if it is False we assume something bad happened and scrap the folder.